src/compute-client/src/controller/instance.rs

```diff
@@ -1536,6 +1557,30 @@ where
         ))
     }
 
+    fn handle_copy_to_response(
+        &mut self,
+        sink_id: GlobalId,
+        response: CopyToResponse,
+    ) -> Option<ComputeControllerResponse<T>> {
+        // We might not be tracking this COPY TO because we have already returned a response
+        // from one of the replicas. In that case, we ignore the response.
+        if self.compute.copy_tos.remove(&sink_id) {
+            let result = match response {
+                CopyToResponse::RowCount(count) => Ok(count),
+                CopyToResponse::Error(error) => Err(anyhow::anyhow!(error)),
+                // We do not expect to receive a `Dropped` message here.
+                // A sink is dropped when a response is already received.
+                // And since we can receive only one response, the `Dropped`
+                // never reaches here.
+                // TODO (mouli): should we panic! instead?
+                CopyToResponse::Dropped => Err(anyhow::anyhow!("unexpected error")),
+            };
+            Some(ComputeControllerResponse::CopyToResponse(sink_id, result))
```

Comment:
@benesch Doing the `CopyToResponse` to `Result` conversion here.

Reply:
Nice. This seems slightly cleaner to me since the weirdness around `CopyToResponse::Dropped` is handled all internally to the controller, and the adapter never sees any of the weirdness. But I defer to @teskje for the final assessment, and also for answering the question in your TODO!

Reply:
The previous version where `Dropped` responses were reported to the caller had the property of being consistent with subscribe handling, so a lot of subscribe code could just be copied and adapted. I don't mind simplifying things, but then we should make extra sure that adapter also handles this response protocol correctly.

For adapter to be able to handle `CopyToResponse`s correctly, we need to agree on what the controller promises about them. I'd suggest the following:

1. For each installed copy-to, the controller will produce _at most_ one response.
2. Once a copy-to was dropped, the controller will not produce responses for it anymore.

With this contract, we'd need to change only one things: In `drop_collections`, remove all dropped IDs from `copy_tos`. Doing this guarantees that we don't return more responses for dropped copy-to. It also makes erroring on a `Dropped` response in `handle_copy_to_response` correct.

Reply:
And in adapter, we should make sure we don't leak state associated with the copy-to when it is canceled before it completes. I think the current adapter-side handling is modeled after subscribe, and while I don't fully grok adapter's subscribe handling, I think it keeps around subscribe state after the `drop_collections` call until it receives a final response?

Reply:
Btw, I had added supporting cancellation as a public preview blocker as a few days ago. The ideal case is that it's easy to rig up now, but I also think it's okay if we don't get the cancellation semantics _quite_ right while we're still iterating. (Memory leaks are a bit of a gray area though; if we release to private preview with a way for users to leak memory in `envd`, that can put the stability of the whole environment at risk.)

![image](https://github.com/MaterializeInc/materialize/assets/882976/447a7fac-6adf-44a2-8b5c-85f26197a4c3)

Reply:
Changed it to match the updated subscribe (returning `None` instead of `Err` for the `Dropped` case). This block is within a `if self.compute.copy_tos.remove(&sink_id) {`, so the state will be removed as soon as we receive one response.